# Importing Libraries

In [79]:
# Import libraries 
import pandas as pd
import numpy as np
import os

# Original Data- products.csv

In [80]:
path = r'/Users/amritamaniram/Desktop/Instacart Basket Analysis 2023'

In [81]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# Prepared Data- orders_wrangled.csv

In [82]:
df_ords = pd.read_csv(os.path.join(path, '02 Data','Prepared Data','orders_wrangled.csv'))

# Mixed-Type Data

In [83]:
# Create a dataframe

In [84]:
df_test = pd.DataFrame ()

In [85]:
# Create a mixed type column

In [86]:
df_test['mix'] = ['a', 'b', 1, True]

In [87]:
df_test.head ()

mix
0     a
1     b
2     1
3  True

In [88]:
# Check for mixed types

In [89]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [90]:
# Converting datatype from numeric to string 

In [91]:
df_test['mix'] = df_test['mix'].astype('str')

# Finding Missing Values

In [92]:
# Finding missing values

In [93]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [94]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [95]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [96]:
# Checking the rumber of rows in product dataframe

In [97]:
df_prods.shape

(49693, 5)

In [98]:
# Creating a new dataframe for non-missing values

In [99]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [100]:
df_prods_clean.shape

(49677, 5)

In [101]:
# Looking for duplicates in products dataframe

In [102]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [103]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [104]:
# Creating new dataframe without duplicates

In [105]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [106]:
df_prods_clean_no_dups.shape

(49672, 5)

In [107]:
# Exporting new dataframe as products_checked

In [108]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data','products_checked.csv'))

# 4.5 Task

# Q.2

In [109]:
# Investigating data

In [110]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  time_elapsed_since_last_order  
count        3.421083e+06       3.421083e+06                   3.214874e+06  
mean         2.776219e+00       1.345202e+01                   1.111484e+01  
std          2.046829e+00       4.226088e+00                   9.206737e+00  
min          0.000000e+00       0.000000e+00                   0.000000e+00  
25%          1.000000e+00       1.000000e+01                   4.000000e+00  
50%          3.000000e+00       1.300000e+01                   7.000000e+00  
75%          5.000000e+00       1.600000e+01                   1.500000e+01  
max          6.000000e+00       2.300000e+01                   3.000000e+01

Q.2 Answer

time_elapsed_since_last_order has a median of 7. However, the min value is: '0' and the max value is '3'.

In 'order_number' the min and max values is '1'. However, the lower quartile is '2' and upper quartile is '5'.

# Q.3

In [111]:
# Check for mix data-type

In [112]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else :
    print ("None of the columns have mix data")

None of the columns have mix data


# Q.4

In [113]:
# No mixed data-type found

# Q.5 

In [114]:
# Checking for missing values

In [115]:
df_ords.isnull().sum()

Unnamed: 0                            0
order_id                              0
user_id                               0
eval_set                              0
order_number                          0
orders_day_of_week                    0
order_hour_of_day                     0
time_elapsed_since_last_order    206209
dtype: int64

There are 206209 missing values in 'time_elapsed_since_last_order'

# Q.6

In [116]:
# Addressing missing values by creating a new data frame with NaN results 

In [117]:
df_ords_nan = df_ords[df_ords['time_elapsed_since_last_order'].isnull() == True]

In [118]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  time_elapsed_since_last_order  
0                         2                  8                            NaN  
11                        2                 11                            NaN  
26                        1                 14                            NaN  
39                        6                 11                            NaN  
45                        3                 12                            NaN  
...                     ...                ...                            ...  
3420930                   4                 12                            NaN  
3420934                   3                 18                            NaN  
3421002                   6                 19                            NaN  
3421019                   1                 15                            NaN  
3421069                   3                 11                            NaN  

[206209 rows x 8 columns]

It seems as though NaN is correlated with order_number 1. Since that is there first recorded order, it would make sense that there is no available data for 'time_elapsed_since_last_order'.
We will not touch these missing values.

# Q.7

In [120]:
# Checking for duplicates

In [121]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [122]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, time_elapsed_since_last_order]
Index: []

In [123]:
# No duplicate records found.

# Q.8

In [124]:
# Since there were no duplicate records found, no action will be taken.

# Export df_ords as orders_checked

In [126]:
df_ords.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_checked.csv'))

# Export df_prods as products_checked

In [127]:
df_prods.to_csv(os.path.join(path,'02 Data','Prepared Data', 'products_checked.csv'))